### PR TITLE
docs: enhance contributing guidelines to be more accessible

### PR DESCRIPTION
**Pull request by ddneilson**
### What was the problem/requirement? (What/Why)

The contributing and development guidelines in this repository could benefit from an improvement pass to make workflows, expectations, and intentions clearer.

### What was the solution? (How)

Rework/refactor the CONTRIBUTING.md, DEVELOPMENT.md, GitHub issue templates, and pull request template with the goal of improving understandability to those not familiar with the code base, hatch-based workflows, GitHub pull request workflows, and our use of conventional commits.

### What is the impact of this change?

The hope is that the development and contribution process should be much clearer for everyone.

### How was this change tested?

I've exercised the development workflows suggested.

### Did you run the "Job Bundle Output Tests"? If not, why not? If so, paste the test results here.

N/A

### Was this change documented?

It is documentation. ;-)

### Is this a breaking change?

No. No functional code is modified.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -6,16 +8,40 @@
...
+
 ### Is this a breaking change?
 
+A breaking change is one that modifies a public contract in some way or otherwise changes functionality of this application in a way
+that is not backwards compatible. Examples of changes that are breaking include:
+
+1. Adding a new required value to the init-data or run-data of the adaptor;
+2. Deleting or renaming a value in the init-data or run-data of the adaptor; and
+3. Otherwise modifying the interface of the adaptor such that a job submitted with an older version of the Maya submitter plug-in
+   will not work a version of the adaptor that includes your modification.
```
Comment (erico-aws) [inline .github/PULL_REQUEST_TEMPLATE.md]: nit: `work *with* a version of...`

File: README.md [hunk 2]
```diff
@@ -20,46 +26,114 @@ This library requires:
...
+If you are using the [default Queue Environment](default-queue-environment), or an equivalent, to run your jobs, then the adaptor will be
+automatically made available to your job. Otherwise, you will need to install the adaptor.
 
 The adaptor can be installed by the standard python packaging mechanisms:
 ```sh
 $ pip install deadline-cloud-for-maya
 ```
 
-After installation it can then be used as a command line tool:
+After installation it can then you can test that it has been installed properly by running the following as the same
```
Comment (erico-aws) [inline README.md]: I think you might have missed deleting some words here.

File: README.md [hunk 3]
```diff
@@ -20,46 +26,114 @@ This library requires:
...
+### Maya Adaptor
+
+Jobs created by this submitter require this adaptor be installed on your worker hosts, and that both the installed adaptor
+and the Maya executable be available on the PATH of the user that will be running your jobs.
+
+The adaptor application is a command-line Python-based application that enhances the functionality of Maya for running
+within a render farm like Deadline Cloud. Its primary purpose for existing is to add a "sticky rendering" functionality
+where a single process instance of Maya is able to load the scene file and then dynamically be instructed to perform
+desired renders without needing to close and re-launch Maya between them, though it also supplies additional benefit
+such as support for path mapping, and reporting the progress of your render to Deadline Cloud. The alternative
```
Comment (crowecawcaw) [inline README.md]: Consider breaking up or shortening this sentence.

File: README.md (hunk 2 above)
Comment (crowecawcaw) [inline README.md]: Typo

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -1,79 +1,225 @@
...
+
+To run the adaptor you will first need to create two files:
+
+1. An `init-data.yaml` (or `init-data.json`) file that contains the information passed to the adaptor
+   during its initialization phase. The schema for this file can be found at
+   `src/deadline/maya_adaptor/MayaAdaptor/schemas/init_data.schema.json`, and examples of init data can
+   be found in the `template.yaml` files spread throughout this repository.
+2. A `run-data.yaml` (or `run-data.json`) file that contains the information passed to the adaptor 
+   to do a single Task run. The schema for this file can be found at
+   `src/deadline/maya_adaptor/MayaAdaptor/schemas/run_data.schema.json`, and examples of init data can
```
Comment (erico-aws) [inline DEVELOPMENT.md]: nit: `examples of *run* data can..`

File: README.md [hunk 5]
```diff
@@ -20,46 +26,114 @@ This library requires:
...
 
-## Versioning
+### Maya Software Availability in AWS Deadline Cloud Service Managed Fleets
 
-This package's version follows [Semantic Versioning 2.0](https://semver.org/), but is still considered to be in its 
-initial development, thus backwards incompatible versions are denoted by minor version bumps. To help illustrate how
-versions will increment during this initial development stage, they are described below:
+You will need to ensure that the version of Maya that you want to run is available on the worker host when you are using
+AWS Deadline Cloud's [Service Managed Fleets](service-managed-fleets) to run jobs;
+hosts do not have any rendering applications pre-installed. The standard way of accomplishing this is described
```
Comment (crowecawcaw) [inline README.md:115]: This is a bit misleading; software isn't preinstalled but their are available conda packages that the submitter's jobs will automatically pull in. (Right? Maybe we don't have a Maya conda package.)

File: README.md [hunk 6]
```diff
@@ -20,46 +26,114 @@ This library requires:
...
-3. The PATCH version is currently incremented when bug fixes or backwards compatible changes are introduced to the public API. 
+## Modifying Jobs Before Submitting
+
+To submit a job, the submitter first generates a [Job Bundle](job-bundle), and then uses functionality from the
+[Deadline](deadline-cloud-client) package to submit the Job Bundle to your render farm to run. If you need to modify
+the job that will be submitted to your farm, then you can use the "Export Bundle" button in the submitter to export the
+Job Bundle to a location of your choice, modify that bundle, and then use the [Deadline Cloud application](deadline-cloud-client)
+to submit that bundle to your farm.
+
+[job-bundle]: https://docs.aws.amazon.com/deadline-cloud/latest/developerguide/build-job-bundle.html
```
Comment (crowecawcaw) [inline README.md]: Is there a customer question that prompted this section? It seems like if someone is using the submitter, it's because they don't want to write a bundle directly.

File: docs/software_arch.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+Fundamentally, what this submitter is doing is creating a [Job Bundle](https://docs.aws.amazon.com/deadline-cloud/latest/developerguide/build-job-bundle.html)
+and using the GUI creation code in the [`deadline` Python package](https://pypi.org/project/deadline/) to generate the UI
+that is displayed to the user. The important parts to know about in a job bundle are:
+
+1. The job template file. The submitter code dynamically generates the template based on properties of the specific scene file
+   that is loaded. For example, it may contain a Step for each layer of the scene to render.
+   Note: All job template files are currently derived from a standard static job template located at
+   `src/deadline/maya_submitter/default_maya_job_template.yaml`.
+2. Asset references. These are the files that the job, when submitted, will require to be able to run. The submitter contains code
+   that introspects the scene that is loaded to automatically discover these, and the submitter UI allows the end-user to modify
```
Comment (crowecawcaw) [inline docs/software_arch.md]: I don't quite follow the second sentence. Can we simplify it?

File: docs/software_arch.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+See the [README](../README.md#maya-adaptor) for background on what purpose the adaptor application serves.
+
+The implementation of the adaptor for Maya has two parts:
+
+1. The adaptor application itself whose code is located in `src/deadline/maya_adaptor/MayaAdaptor`. This is the
+   implementation of the command-line application (named `maya-openjd`) that is run by Jobs created by the Maya submitter.
+2. A "MayaClient" application located in `src/deadline/maya_adaptor/MayaClient`. This is an application that is
+   run within Maya itself by the adaptor application when it launches Maya. The MayaClient facilitates communication
+   between the adaptor itself and the running Maya application; communication to tell Maya to, say, load a scene file,
+   or render frame 20 of the loaded scene.
```
Comment (crowecawcaw) [inline docs/software_arch.md]: I think one important detail in here is that the client starts with Maya and runs for the duration of the session, while the adapter runs once per task.

Also a nit: many uses of "itself" which can be removed.

File: docs/software_arch.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+The adaptor application itself is built using the [Open Job Description Adaptor Runtime](https://github.com/OpenJobDescription/openjd-adaptor-runtime-for-python)
+package. This package supplies the application entrypoint that defines and parses the command-line subcommands and options, as well as
+the business logic that drives the state machine of the adaptor itself. Please see the README for the runtime package for information on
+the lifecycle states of an adaptor, and the command line options that are available. 
+
+Digging through the code for the adaptor, you will find that the `MayaAdaptor.on_start()` method is where the Maya application is started.
+Rather than starting it with command-line options that say to render a specific scene, the application is started with arguments that
```
Comment (crowecawcaw) [inline docs/software_arch.md]: Consider removing the first clause "Rather than..." so the sentence just says what the application does instead of comparing it another implicit design.

File: docs/software_arch.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+Digging through the code for the adaptor, you will find that the `MayaAdaptor.on_start()` method is where the Maya application is started.
+Rather than starting it with command-line options that say to render a specific scene, the application is started with arguments that 
+tell Maya to run the "MayaClient" application. This application is, essentially, a secure web server that is running over named pipes
+rather than network sockets. The adaptor sends the client commands (look for calls to `enqueue_action()` in the adaptor) to instruct Maya
+to do things, and then waits for the results of those actions to take effect. 
+
+You can see the definitions of the available commands, and the actions that they take by inspecting `src/deadline/MayaClient/maya_client.py`. You'll
+notice that the commands that it directly defines are minimal, and that the set of commands that are available is updated when the adaptor sends
+it a command to set the rendererer being used. Each renderer has its own command handler defined under `src/deadline/MayaClient/render_handlers`.
```
Comment (crowecawcaw) [inline docs/software_arch.md]: Typo: rendererer

**Review by crowecawcaw — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (ddneilson) [reply README.md:115]: Right, software isn't preinstalled. That's what I meant by "hosts do not have any rendering applications pre-installed". The links in the subsequent sentences point the customer to the documentation on using conda to install apps, and the ones that we make available in the service's internal conda channel. 

I will add "... hosts do not have any rendering applications pre-installed" to the end of the last sentence in the paragraph, though, to help make things a bit clearer. Does that help?

File: README.md (hunk 6 above)
Comment (ddneilson) [reply README.md]: Yes, actually. I know of at least one internal customer that wanted to modify the bundle but wasn't aware of the 'Export bundle' function. The intent of this section is to call attention to that functionality.

Though, I can rephrase this to be more around seeing what will be submitted.

File: docs/software_arch.md (hunk 7 above)
Comment (ddneilson) [reply docs/software_arch.md]: Hah. Yeah, for sure. I wrote it and I can't even make sense of the sentence :-)

File: docs/software_arch.md (hunk 8 above)
Comment (ddneilson) [reply docs/software_arch.md]: reworded.

**Review by erico-aws — DISMISSED** [review]

**Review by crowecawcaw — DISMISSED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 11]
```diff
@@ -6,16 +12,48 @@
...
+*delete text starting here*
+A breaking change is one that modifies a public contract in some way or otherwise changes functionality of this application in a way
+that is not backwards compatible. Examples of changes that are breaking include:
+
+1. Adding a new required value to the init-data or run-data of the adaptor;
+2. Deleting or renaming a value in the init-data or run-data of the adaptor; and
+3. Otherwise modifying the interface of the adaptor such that a job submitted with an older version of the Maya submitter plug-in
+   will not work with a version of the adaptor that includes your modification.
+
+If so, then please describe the changes that users of this package must make to update their scripts, or Python applications. Also,
```
Comment (jairaws) [inline .github/PULL_REQUEST_TEMPLATE.md:52]: Should breaking changes be published in a new major version? How do we control versioning?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -6,62 +6,115 @@ documentation, we greatly value feedback and contributions from our community.
...
+* `fix` - When the pull request is implementing a fix to a bug;
+* `test` - When the pull request is only implementing an addition or change to tests or the testing infrastructure;
+* `docs` - When the pull request is primarily implementing an addition or change to the package's documentation;
+* `refactor` - When the pull request is implementing only a refactor of existing code;
+* `ci` - When the pull request is implementing a change to the CI infrastructure of the packge;
+* `chore` - When the pull request is a generic maintenance task.
 
-## Security issue notifications
+We also require that the type in your conventional commit title end in an exclaimation point (e.g. `feat!` or `fix!`)
+if the pull request should be considered to be a breaking change in some way. Please also include a "BREAKING CHANGE" footer
```
Comment (jairaws) [inline CONTRIBUTING.md:110]: What are the consequences creating a PR with breaking changes that are not tagged as so?

**Review by jairaws — APPROVED** [review]
Great improvements to the documentation! Thanks!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 11 above)
Comment (ddneilson) [reply .github/PULL_REQUEST_TEMPLATE.md:52]: This is the [versioning](https://github.com/aws-deadline/deadline-cloud-for-maya/?tab=readme-ov-file#versioning) section of the README.
While we're at major version 0, minor version changes indicate breaking changes.

File: CONTRIBUTING.md (hunk 12 above)
Comment (ddneilson) [reply CONTRIBUTING.md:110]: We release a version that increments the patch version rather than the minor version. This could cause problems for consumers. e.g. Say they submit a job with the new submitter (v0.20.5) that requires a breaking change that was made in an adaptor. If we only increment the patch level then it looks like any v0.20.x version of the adaptor will be able to run it, but actually only 0.20.5 will be able to.

**Review by crowecawcaw — APPROVED** [review]

**Review by leongdl — COMMENTED** [review]
Thank you for the updates! Great to have more contributions.
